File: examples/tlplay-python/PlaybackBar.py
# SPDX-License-Identifier: BSD-3-Clause
# Copyright Contributors to the feather-tk project.

import opentimelineio as otio
import ftkPy as ftk
import tlRenderPy as tl

import PlaybackActions

class Widget(ftk.IWidget):

    def __init__(self, context, app, actions, parent = None):
        ftk.IWidget.__init__(self, context, "PlaybackBar.Widget", parent)

        self._player = None

        self._playbackToolBar = ftk.ToolBar(context)
        self._playbackToolBar.addAction(actions.actions["Reverse"])
        self._playbackToolBar.addAction(actions.actions["Stop"])
        self._playbackToolBar.addAction(actions.actions["Forward"])

        self._frameToolBar = ftk.ToolBar(context)
        self._frameToolBar.addAction(actions.actions["Start"])
        button = self._frameToolBar.addAction(actions.actions["Prev"])
        button.repeatClick = True
        button = self._frameToolBar.addAction(actions.actions["Next"])
        button.repeatClick = True
        self._frameToolBar.addAction(actions.actions["End"])

        self._currentTimeEdit = tl.ui.TimeEdit(context, app.getTimeUnitsModel())
        self._currentTimeEdit.tooltip = "Current time."

        self._durationLabel = tl.ui.TimeLabel(context, app.getTimeUnitsModel())
        self._durationLabel.tooltip = "Playback duration."

        self._speedEdit = ftk.DoubleEdit(context)
        self._speedEdit.range = ftk.RangeD(1.0, 99999.0)
        self._speedEdit.step = 1.0
        self._speedEdit.largeStep = 10.0
        self._speedEdit.tooltip = "Playback speed."

        self._timeUnitsWidget = tl.ui.TimeUnitsWidget(context, app.getTimeUnitsModel())
        self._timeUnitsWidget.tooltip = "Time units."

        self._layout = ftk.HorizontalLayout(context, self)
        self._layout.marginRole = ftk.SizeRole.MarginInside
        self._playbackToolBar.parent = self._layout
        self._frameToolBar.parent = self._layout
        self._currentTimeEdit.parent = self._layout
        self._durationLabel.parent = self._layout
        self._speedEdit.parent = self._layout
        self._timeUnitsWidget.parent = self._layout
        
        self._currentTimeEdit.setCallback(self._currentTimeCallback)

        self._speedEdit.setCallback(self._speedCallback)

        self._playerObserver = tl.timeline.ValueObserverPlayer(
            app.getDocumentModel().observePlayer(),
            self._widgetUpdate)

    def setGeometry(self, value):
        ftk.IWidget.setGeometry(self, value)
        self._layout.setGeometry(value)
    
    def sizeHintEvent(self, event):
        self.setSizeHint(self._layout.sizeHint)

    def _currentTimeCallback(self, value):
        if self._player:
            self._player.currentTime = value

    def _currentTimeUpdate(self, value):
        self._currentTimeEdit.value = value

    def _speedCallback(self, value):
        if self._player:
            self._player.speed = value

    def _speedUpdate(self, value):
        self._speedEdit.value = value

    def _widgetUpdate(self, player):
        self._player = player
        if player:
            self._durationLabel.value = player.duration
            self._currentTimeObserver = tl.timeline.ValueObserverRationalTime(
                player.observeCurrentTime,
                self._currentTimeUpdate)
            self._speedObserver = ftk.ValueObserverD(
                player.observeSpeed,
                self._speedUpdate)
        else:
            self._currentTimeEdit.value = tl.invalidTime
            self._durationLabel.value = tl.invalidTime
            self._speedEdit.value = 1.0
            self._currentTimeObserver = None
            self._speedObserver = None
        self._playbackToolBar.enabled = player != None
        self._frameToolBar.enabled = player != None
        self._currentTimeEdit.enabled = player != None
        self._durationLabel.enabled = player != None
        self._speedEdit.enabled = player != None

File: lib/tlRender/PlayApp/PlaybackBar.cpp
// SPDX-License-Identifier: BSD-3-Clause
// Copyright Contributors to the tlRender project.

#include "PlaybackBar.h"

#include "App.h"
#include "FilesModel.h"

#include <ftk/UI/ToolButton.h>
#include <ftk/Core/Format.h>

namespace tl
{
    namespace play
    {
        void PlaybackBar::_init(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::map<std::string, std::shared_ptr<ftk::Action> >& actions,
            const std::shared_ptr<IWidget>& parent)
        {
            IWidget::_init(context, "PlaybackBar", parent);

            _layout = ftk::HorizontalLayout::create(context, shared_from_this());
            _layout->setMarginRole(ftk::SizeRole::MarginInside);

            auto hLayout = ftk::HorizontalLayout::create(context, _layout);
            hLayout->setSpacingRole(ftk::SizeRole::None);
            auto tmp = actions;
            auto reverseButton = ftk::ToolButton::create(context, tmp["Reverse"], hLayout);
            auto stopButton = ftk::ToolButton::create(context, tmp["Stop"], hLayout);
            auto forwardButton = ftk::ToolButton::create(context, tmp["Forward"], hLayout);
            _loopWidget = ui::PlaybackLoopWidget::create(context, hLayout);
            _loopWidget->setTooltip("Playback loop mode.");

            hLayout = ftk::HorizontalLayout::create(context, _layout);
            hLayout->setSpacingRole(ftk::SizeRole::None);
            auto startButton = ftk::ToolButton::create(context, tmp["Start"], hLayout);
            auto prevButton = ftk::ToolButton::create(context, tmp["Prev"], hLayout);
            prevButton->setRepeatClick(true);
            auto nextButton = ftk::ToolButton::create(context, tmp["Next"], hLayout);
            nextButton->setRepeatClick(true);
            auto endButton = ftk::ToolButton::create(context, tmp["End"], hLayout);

            auto timeUnitsModel = app->getTimeUnitsModel();
            _currentTimeEdit = ui::TimeEdit::create(context, timeUnitsModel, _layout);
            _currentTimeEdit->setTooltip("The current time.");

            _durationLabel = ui::TimeLabel::create(context, timeUnitsModel, _layout);
            _durationLabel->setTooltip("The timeline duration.");

            _speedEdit = ftk::DoubleEdit::create(context, _layout);
            _speedEdit->setRange(ftk::RangeD(1.0, 99999.0));
            _speedEdit->setStep(1.0);
            _speedEdit->setLargeStep(10.0);
            _speedEdit->setTooltip("The timeline speed.");

            _speedMultLabel = ftk::Label::create(context, _layout);
            _speedMultLabel->setHMarginRole(ftk::SizeRole::MarginInside);
            _speedMultLabel->setTooltip("Playback speed multiplier.");

            _timeUnitsWidget = ui::TimeUnitsWidget::create(context, timeUnitsModel, _layout);
            _timeUnitsWidget->setTooltip("Set the time units.");

            _loopWidget->setCallback(
                [this](Loop value)
                {
                    if (_player)
                    {
                        _player->setLoop(value);
                    }
                });

            _currentTimeEdit->setCallback(
                [this](const OTIO_NS::RationalTime& value)
                {
                    if (_player)
                    {
                        _player->stop();
                        _player->seek(value);
                    }
                });

            _speedEdit->setCallback(
                [this](double value)
                {
                    if (_player)
                    {
                        _player->setSpeed(value);
                    }
                });

            _playerObserver = ftk::Observer<std::shared_ptr<Player> >::create(
                app->getFilesModel()->observePlayer(),
                [this](const std::shared_ptr<Player>& value)
                {
                    _player = value;

                    if (value)
                    {
                        _durationLabel->setValue(value->getTimeRange().duration());

                        _loopObserver = ftk::Observer<Loop>::create(
                            value->observeLoop(),
                            [this](Loop value)
                            {
                                _loopWidget->setLoop(value);
                            });

                        _currentTimeObserver = ftk::Observer<OTIO_NS::RationalTime>::create(
                            value->observeCurrentTime(),
                            [this](const OTIO_NS::RationalTime& value)
                            {
                                _currentTimeEdit->setValue(value);
                            });

                        _speedObserver = ftk::Observer<double>::create(
                            value->observeSpeed(),
                            [this](double value)
                            {
                                _speedEdit->setValue(value);
                            });

                        _speedMultObserver = ftk::Observer<double>::create(
                            value->observeSpeedMult(),
                            [this](double value)
                            {
                                _speedMultLabel->setText(ftk::Format("{0}X").arg(value, 1));
                                _speedMultLabel->setBackgroundRole(value > 1.0 ?
                                    ftk::ColorRole::Checked :
                                    ftk::ColorRole::None);
                            });
                    }
                    else
                    {
                        _loopWidget->setLoop(Loop::Loop);
                        _currentTimeEdit->setValue(invalidTime);
                        _durationLabel->setValue(invalidTime);
                        _speedMultLabel->setText("1X");
                        _speedMultLabel->setBackgroundRole(ftk::ColorRole::None);

                        _loopObserver.reset();
                        _currentTimeObserver.reset();
                        _speedObserver.reset();
                        _speedMultObserver.reset();
                    }

                    _loopWidget->setEnabled(value.get());
                    _currentTimeEdit->setEnabled(value.get());
                    _durationLabel->setEnabled(value.get());
                    _speedEdit->setEnabled(value.get());
                    _speedMultLabel->setEnabled(value.get());
                });
        }

        PlaybackBar::~PlaybackBar()
        {}

        std::shared_ptr<PlaybackBar> PlaybackBar::create(
            const std::shared_ptr<ftk::Context>& context,
            const std::shared_ptr<App>& app,
            const std::map<std::string, std::shared_ptr<ftk::Action> >& actions,
            const std::shared_ptr<IWidget>& parent)
        {
            auto out = std::shared_ptr<PlaybackBar>(new PlaybackBar);
            out->_init(context, app, actions, parent);
            return out;
        }

        ftk::Size2I PlaybackBar::getSizeHint() const
        {
            return _layout->getSizeHint();
        }

        void PlaybackBar::setGeometry(const ftk::Box2I& value)
        {
            IWidget::setGeometry(value);
            _layout->setGeometry(value);
        }
    }
}